e2e-testing/readme.md

```diff
@@ -0,0 +1,164 @@
+
+| Unit Test | Integration Test | E2E Test |
+|-----------|------------------|----------|
+| A module, APIs | Modules, Interfaces | Application, System |
+| Tiny | Small to Medium | Large |
+| Transparent | Black box | Black box |
+| Isolated | Test interfaces and flow data between the modules | Application flow from start to end
+
+The main reason for carrying out E2E testing is to determine various dependencies of an application as well as ensuring that accurate information is communicated between various system components. It is usually performed after the completion of functional and system testing of any application.
+
+## Why Do We Perform E2E Testing
+
+In many commercial software application scenarios, a modern software system comprises of its interconnection with multiple sub-systems. These sub-systems can be within the same organization or can be components of different organizations. Also, these sub-systems can have somewhat similar or different lifetime release cycle from the current system. As a result, if there is any failure or fault in any sub-system, it can adversely affect the whole software system leading to its collapse.
+
+![E2E Testing Pyramid](./images/e2e-pyramid.png)
+
+The above illustration is a testing pyramid from [Kent C. Dodd’s blog](https://blog.kentcdodds.com/write-tests-not-too-many-mostly-integration-5e8c7fff591c) which is a combination of the pyramids from [Martin Fowler’s blog](https://martinfowler.com/bliki/TestPyramid.html) and the [Google Testing Blog](https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html).
+
+The majority of your tests are at the bottom of the pyramid. As you move up the pyramid, the number of tests gets smaller. Also, going up the pyramid, tests get slower and more expensive to write, run, and maintain. Therefore, before delving yourself in E2E testing specifications and planning, it is important to keep the checklist of the ***essential activities that should be included in every E2E process***:
+
+- A thorough study of requirements to perform E2E testing.
+- Proper set up of test environments.
+- A thorough study of Hardware and Software requirements.
+- Descriptions of all subsystems as well as main software system involved.
+- Enlist the roles and responsibilities for all the systems and subsystems involved.
+- Testing methods used under E2E testing as well as standards that are followed, its description.
+- Test cases designing as well as tracing requirement matrix.
+- Record or save the input and output data for each system.
+
+When working on large and complex customer engagements, it is very important to perform E2E testing in a controlled environment, and for a good reason. Even  slight change in the sub-system implementation, the interface contract, data schema, etc. can cause interface conflicts. Therefore, we need **constantly** and **automatically** control E2E tests to manage the scope and breadth of the conflicts.
+
+## E2E Testing Methods
```

Comment:
Can you talk about input and expected output? You mention in the table above this being a black box. How do you see into that black box for failures? (I would expect the existing telemetry tooling should give you _insights_)

Reply:
I have revised and re-written this complete section.